### PR TITLE
New banning misbehavior handling and Peers Tab on Debug Console

### PR DESCRIPTION
**Pull request by jamescowens**
The design is a sparse map keyed by CAddress, where no entry means a misbehavior of zero.
If a node is banned, the misbehavior entry is left in place. It is erased (meaning zeroed...) when the node is unbanned, or the banlist is cleared. This is done by match to sub_net, because manual bans can be placed on individual addresses or entire subnets.  Automatic bans are always on individual addresses (subnets with /32 mask) and misbehaviors are also on individual addresses and the service port number (CAddress).

I have debated whether the banlist should be indexed by CNetAddr rather than CAddress, and have decided CAddress (which includes the remote port number) is the right one. Someone could have several nodes running on the same address through port forwarding/NAT, and if we use CNetAddr only, if one node misbhehaves, it would result in banning all nodes behind the address. The disadvantage of CAddress is that a node that disconnects and reconnects using different port numbers can evade banning for moderate misbehavior by spreading out the misbehavior to multiple CAddress entries corresponding to different ports with the same IP.

misbehavior also decays linearly by default banscore (100) / 86400 every second.

This also implements the peers tab on the debug console, which allows GUI users to interact with the nodes and ban/unban by right-clicking.

### CODE REVIEW

**Comment by jamescowens** [comment]
Note that I have tested this on testnet and a crippled node on mainnet with good results. The DoS BOOST unit tests have also been expanded to cover the new functionality.

File: src/banman.cpp [hunk 1]
```diff
@@ -217,3 +230,35 @@ void BanMan::SetBannedSetDirty(bool dirty)
...
     m_is_dirty = dirty;
 }
+
+unsigned int BanMan::ZeroMisbehavior(CNetAddr net_addr)
+{
+    CSubNet sub_net(net_addr);
+    return ZeroMisbehavior(sub_net);
+}
+
+unsigned int BanMan::ZeroMisbehavior(CSubNet sub_net)
```
Comment (cyrossignol) [inline src/banman.cpp:240]: I think we can put the `ZeroMisbehavior()` methods on `CNode`. Written that way, we can remove the `friend` visibility exposure that it looks like we don't really need here.

**Review by cyrossignol — COMMENTED** [review]
This looks great so far. Will do some more in-depth testing tomorrow.

File: src/banman.cpp (hunk 1 above)
Comment (jamescowens) [reply src/banman.cpp:240]: Good point.

Comment (jamescowens) [reply src/banman.cpp:240]: They will be static, because they are not exactly actions taken on a specific node object per se, but that is ok.

**Comment by jamescowens** [comment]
I wonder if the default decay rate is too rapid? I should probably put a tunable in there for the rate, with a default value.

File: src/banman.cpp (hunk 1 above)
Comment (jamescowens) [reply src/banman.cpp:240]: After looking at this I think I disagree with you. I would rather leave them where they are and make them private in banman. They are only consumed by banman, and in fact putting them in CNode, we would have to make them public to be accessible by banman, or protected, which again requires the friendship.

Making them public to eliminate the friend requirement then makes zeroing out the misbehavior accessible by any class, where having them private to banman only allows banman access. It makes sense to expose CNode protected structures to banman, as that is a relatively narrow surface area, IMHO.

**Comment by iFoggz** [comment]
Would like to see this for more testing and testing on a heavy node.

File: src/banman.cpp (hunk 1 above)
Comment (cyrossignol) [reply src/banman.cpp:240]: I think you're right about that after all. I was thinking about it in terms of unit testing and keeping `BanMan` out of `CNode`'s internals, but it looks like it makes sense in this case. Let's leave it as is.

Comment (jamescowens) [reply src/banman.cpp:240]: Cool.

**Comment by jamescowens** [comment]
The latest force push looks really good to me. Compiles and runs well on all my devices...
OpenSuSE x86_64
Ubuntu 18.04
Windows 64-bit and 32-bit
Ubuntu 18.04 Odroid XU4 (ARM)

I don't have a Mac to check that.

I am running on my scraper node for mainnet with good results.

**Comment by jamescowens** [comment]
Here is a screenshot of this running on my Windows testnet node...
image
Notice there is an IPv6 address in there!

**Comment by iFoggz** [comment]
yup ipv6 is a growing in usage. my big node has ipv6 as well. ipv6 also reduces nat issues and also ipv6 is preferred over ipv4 if both available.

File: src/net.cpp [hunk 2]
```diff
@@ -1232,7 +1287,7 @@ void ThreadSocketHandler2(void* parg)
             //
             // Allow newbies to connect easily
             int64_t nTime = GetAdjustedTime();
-            if (nTime - pnode->nTimeConnected > 24)
+            if (nTime - pnode->nTimeConnected > 45)
```
Comment (cyrossignol) [inline src/net.cpp]: Just curious... what's the reason for this change?

**Review by cyrossignol — APPROVED** [review]
Seems like everything is working and the new tab looks great :slightly_smiling_face:

File: src/net.cpp (hunk 2 above)
Comment (jamescowens) [reply src/net.cpp]: due to our older net code, bignodes sometimes do not send a version message within that period of connecting, so I upped it. #1548 optimizes this portion of net processing and actually gets rid of that setting altogether.

**Review by iFoggz — APPROVED** [review]

**Comment by jamescowens** [comment]
This is to prevent conflicts with #1538.
